source/opt/ilist.h

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_ILIST_H_
+#define LIBSPIRV_OPT_ILIST_H_
+
+#include <cassert>
+#include <type_traits>
+
+#include "ilist_node.h"
+
+namespace spvtools {
+namespace utils {
+
+// An IntrusiveList is a generic implementation of a doubly-linked list.  The
+// intended convention for using this container is:
+//
+//      class Node : public IntrusiveNodeBase<Node> {
+//        // Node that "Node", the class being defined is the template.
+//        // Must have a default constructor accessible to List.
+//        // Add whatever data is needed in the node
+//      };
+//
+//      typedef IntrusiveList<Node> List;
+//
+// You can also inherit from IntrusiveList instead of a typedef if you want to
+// add more functionality.
+//
+// The condition on the template for IntrusiveNodeBase is there to add some type
+// checking to the container.  The compiler will still allow inserting elements
+// of type IntrusiveNodeBase<Node>, but that would be an error. This assumption
+// allows NextNode and PreviousNode to return pointers to Node, and casting will
+// not be required by the user.
+
+template <class NodeType>
+class IntrusiveList {
+ public:
+  static_assert(
+      std::is_base_of<IntrusiveNodeBase<NodeType>, NodeType>::value,
+      "The type from the node must be derived from IntrusiveNodeBase, with "
+      "itself in the template.");
+
+  // Creates an empty list.
+  inline IntrusiveList();
+
+  // Moves the contents of the given list to the list being constructed.
+  IntrusiveList(IntrusiveList&&);
+
+  // Destorys the list.  Note that the elements of the list will not be deleted,
+  // but thy will be removed from the list.
+  ~IntrusiveList();
+
+  // Moves all of the elements in the list on the RHS to the list on the LHS.
+  IntrusiveList& operator=(IntrusiveList&&);
+
+  // Basetype for iterators so an IntrusiveList can be traversed like STL
+  // containers.
+  template <class T>
+  class iterator_template {
+   public:
+    inline iterator_template(const iterator_template& i) : node_(i.node_) {}
+    inline iterator_template& operator++() {
```

Comment:
This method is invalid to call on an end iterator.  If that's intentional then you should have a method comment saying so.
Otherwise guard the assignment, and have a method comment saying what it does on an end iterator.

I see it's useful to have an efficiency gain when you avoid the guard.  I could go either way.

Reply:
When I was thinking of operator++ and operator--, it does not make sense to increment end().  At the same time, I was think that we could add reverse iterators (with rbegin and rend) if we want to traverse a list in reverse.  However, looking on stackoverflow, it seems like people will decrement end().  I'll make that legal.

I'll make end() point to the sentinel node, but this means I can't use PreviousNode() and NexNode() to move the iterator.